### ISSUE TITLE
[bug #43626] Inconsistent timeouts in stk500v2

### ISSUE THREAD

**Issue by avrdude-admin** [avrdudes/avrdude#362]
Matthijs Kooijman <None>
Mon 17 Nov 2014 04:42:50 PM UTC

While investigating a problem, I noticed the timeouts for stk500v2_recv are a bit inconsistent. In particular, ser_posix defines:

long serial_recv_timeout = 5000; /* ms */

While stk500v2.c defines:

#define SERIAL_TIMEOUT 2

In practice, I think this means that when no data is being received, serial_recv blocks for 5 seconds. However, when a byte (that is not the final byte in the packet) **is** received, > 2 seconds after stk500v2_recv was called, it returns a timeout immediately.

Effectively this means that there's 3 seconds spent waiting for data this won't be processed anyway (unless it's the last byte).

Additionally, it seems that the 5 second timeout coincides with the timeout specified by AVR068 for "data" packets and it seems useful to let avrdude timeout (and thus retry) before the stk500 times out.

> The total timeout period is from a command is sent to the answer must be
> completely received. The total timeout period is 200 ms for the
> CMD_SIGN_ON command, 5 seconds for the CMD_READ/PROGRAM_FLASH/EEPROM
> commands, and 1 second for all other commands.

This issue was migrated from https://savannah.nongnu.org/bugs/?43626

**Comment by avrdude-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Wed 19 Nov 2014 04:47:38 PM UTC

I can see what you mean, but hesitate to change something
there, as it requires a lot of testing against a number of
possible programmer implementations.

Do you have any indication this situation (character received
by ser_posix, but then thrown away due to being older than
two seconds) is really triggered in any practical situation,
or is this rather a theoretical possibility?

**Comment by avrdude-admin** [comment]
Matthijs Kooijman <matthijs>
Wed 19 Nov 2014 05:07:55 PM UTC

> Do you have any indication this situation (character received

by ser_posix, but then thrown away due to being older than
two seconds) is really triggered in any practical situation,
or is this rather a theoretical possibility?

No, it's just something I noticed while looking at the code. I did have some timeout-related problems, but I don't think this actually influences it.

I don't mind to keep it like this now, though perhaps adding a comment here or there might help prevent future confusion (people thinking there is a retry after 2s while it's really after 5s or something)? Just a suggestion :-)
